### PR TITLE
Upgrade to Fastify v3

### PR DESCRIPTION
**Pull request by thomheymann**
This PR upgrades `fastify-plugin` to Fastify v3. The only changes were TypeScript related but I also aligned the filenames so that implementation and declaration files are consistent.

I have followed the new TypeScript guide here: https://github.com/fastify/fastify/pull/1966/files?short_path=17177d7#diff-17177d7eb9ef1a38f00590c051d2b873

However, I am not sure if the `FastifyPlugin` interface is correct as its requires knowledge of server, request and response types of the fastify instance at build time of the plugin. We don't know what the fastify instance looks like at build time though. 

@Ethan-Arrowood - Can you have a look at the `test/types.test.ts`? I have added a test for this case so build currently fails

### Example plugin and usage with new `FastifyPlugin` type
```ts
export const testPlugin: FastifyPlugin<{}> = fp(
  function (fastify, options, _next) {
    fastify.decorate('utility', () => options.hello)
  },
  '>=1'
);

const server = fastify()
server.register(testPlugin) // This works since `FastifyPlugin` defaults to HTTP
```

### Example plugin with HTTP2 Fastify instance
```ts
const serverWithHttp2: FastifyInstance<Http2Server, Http2ServerRequest, Http2ServerResponse> = fastify({ http2: true });

serverWithHttp2.register(testPluginWithOptions) // This fails since `FastifyPlugin` defaulted to HTTP but Fastify instance is HTTP2
```

Yes, we could cast to unknown first and then to HTTP2 version of `FastifyPlugin` but I don't think that's a good type system:

```ts
serverWithHttp2.register((testPluginWithOptions as unknown) as FastifyPlugin<Options, Http2Server, Http2ServerRequest, Http2ServerResponse>) // 🙈
```

Might be missing something, let me know how you want this to work. 

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by mcollina** [comment]
Would you mind to include those two examples in the README?

**Comment by mcollina** [comment]
Also, CI is failing everywhere.

Plus, can you drop Node 6 and 8? Thanks.

**Comment by Ethan-Arrowood** [comment]
@thomheymann are you aware that this work is already done in this PR here: https://github.com/fastify/fastify-plugin/pull/81?

I'm happy to let you contribute to that branch if you'd like to complete the PR but I believe I was waiting to publish it when the first v3 release candidate was released

**Comment by thomheymann** [comment]
Hi Ethan, I wasn't aware of the PR #81 But I think you're going to run into the same typing issues with HTTP2 Fastify instances.

**Comment by Ethan-Arrowood** [comment]
Alright cool I'm in favor of this pr then. I'll do a proper review now 😄

File: test/types.test.ts [hunk 1]
```diff
@@ -31,7 +36,20 @@ export const testPluginWithAsync = fp(
...
 const server = fastify()
 
-server.register(testPlugin) // register expects a fastify.Plugin
-server.register(testPluginWithCallback) // test fastify.Plugin's gerneric typings
+server.register(testPluginWithOptions) // register expects a FastifyPlugin
+server.register(testPluginWithCallback)
+server.register(testPluginWithAsync)
+
+// Register with HTTP2
+const serverWithHttp2: FastifyInstance<
```
Comment (Ethan-Arrowood) [inline test/types.test.ts]: You shouldn't need to specify these generics any longer. Let me know if removing them causes issues and I'll jump in and try and figure out whats up.

File: tsconfig.json [hunk 2]
```diff
@@ -4,9 +4,9 @@
     "module": "commonjs",
     "noEmit": true,
     "strict": true,
-    "noImplicitAny": false,
+    "noImplicitAny": true,
```
Comment (Ethan-Arrowood) [inline tsconfig.json]: I think this is true by default with `strict: true`? try removing it and see if stuff still works.

File: test/types.test.ts (hunk 1 above)
Comment (thomheymann) [reply test/types.test.ts]: I don't need to specify the generics but if I remove them I get HTTP types, not HTTP2 types.

I think we might need to look at the types themselves. 

When authoring a plugin, we will not know what generics will be passed into `FastifyInstance` so maybe the types need to cater for that? 
e.g. `FastifyPlugin<Server = http.Server & https.Server & http2.Http2Server>`

The other issue I find strange is that server type is determined by fastify options. If https options are passed in, the server is of type https. If http2 boolean is passed in, the server is of that type. Shouldn't the types reflect that mechanism?

Comment (Ethan-Arrowood) [reply test/types.test.ts]: Not exactly, the server types are *supposed* to be inferred based on the options object. The generics only exist if you want to extend the types of those servers, request, or responses. See here:

https://github.com/fastify/fastify/blob/f1cfcadbe201f98137a5b2a9ee1159215b5d81e5/fastify.d.ts#L20-L44

This is how the `fastify()` method is defined in the `next` types (if you're curious its using discriminant unions and function overloads to achieve it).

Fastify-Plugin should forward the generic properties through to the underlying instance, and it should also enable the user to omit them if they are not decorating those generic parameters

Comment (thomheymann) [reply test/types.test.ts]: You're right, the server type is correctly inferred, that's great!

I'm still not clear how the `FastifyPlugin` type is supposed to be used though.

Plugins are defined like so:

```ts
export const testPlugin: FastifyPlugin = fp(
  function (fastify, _options, _next) {
    console.log(fastify) // Is now of type `FastifyInstance<Server, IncomingMessage, ServerResponse, FastifyLoggerOptions<Server, IncomingMessage, ServerResponse>>`
  },
  {},
)
```

How will TypeScript know what options have been passed into `fastify`? The fastify instance hasn't even been created at that point.

Consequently in user land, the following line is failing because of the defaults used by `FastifyPlugin` type:

```ts
const serverWithHttp2 = fastify({ http2: true });
serverWithHttp2.register(testPlugin) // Argument of type 'FastifyPlugin<{}, Server, IncomingMessage, ServerResponse>' is not assignable to parameter of type 'FastifyPlugin<{}, Http2Server, Http2ServerRequest, Http2ServerResponse>'.ts(2345)
```

Can you explain how to make this work correctly?

Comment (Ethan-Arrowood) [reply test/types.test.ts]: Hmm this is a touch challenge. Let me check this out locally and start investigating.

Comment (Ethan-Arrowood) [reply test/types.test.ts]: @mcollina can you foresee any circumstance where a plugin should only work for a specific type (http vs https vs http2) of fastify server? Disregarding TypeScript, is there anything currently in place that validates that a (for example) http2-only plugin is being registered in a http2 fastify instance?

Comment (Ethan-Arrowood) [reply test/types.test.ts]: @thomheymann after working with this for a little bit I believe we are trying to solve a problem that doesn't exist (_yet_). As far as I'm aware plugins aren't scoped to a specific type of server instance and as such shouldn't have to pass some type check for that either. I was able to resolve the issue you're facing here by removing the defaults in the FastifyPlugin definition in the fastify type defs. This still enforces that the server, request, and response all match, but then allows for any plugin to be registered to any server instance. I'm going to wait Matteo's reply and then will proceed as I have other ideas for solutions depending on what he says.

Comment (mcollina) [reply test/types.test.ts]: @mcollina can you foresee any circumstance where a plugin should only work for a specific type (http vs https vs http2) of fastify server? 

No, the vast majority of plugins should infer the http vs https vs http2 thing automatically. I _thought_ this was what v3 types were about :/.

Comment (thomheymann) [reply test/types.test.ts]: @Ethan-Arrowood  Can you share your solution? When I remove the defaults I get typing errors asking me to provide types for the generics

Comment (Ethan-Arrowood) [reply test/types.test.ts]: Okay I'll work on a response soon. Its sorta complicated. 

Short answer is this:
>  I was able to resolve the issue you're facing here by removing the defaults in the FastifyPlugin definition in the fastify type defs

Without modifying Fastify types you can cast a plugin to `unknown` then cast it to your intended FP instance inline which is a bit better that specifying with generic parameters. Also, the answer might be to just use generic parameter inference. Its not a great solution but it definitely is one. 

So like I said let me spend some more time with this and ill write back as soon as I can!

Comment (Ethan-Arrowood) [reply test/types.test.ts:47]: I'm not finished yet, but I'm starting to simplify the problem in a [TypeScript playground](https://www.typescriptlang.org/play/index.html?ssl=2&ssc=22&pln=1&pc=1#code/JYWwDg9gTgLgBACxjMBnOAzKERwORIqp4BQoksiyYm2uB1p508hYATLTvm+6STACeYAKZwA8mBjAIAO3QBeOAG8ScdXAA2EAOY6RUAPwAuOACMIETSICGskgF8SA4WIAS1VJOlzFEqTLycABkKmoabKimkQB0AMoGAG4G3oGojs5ConAeKOypvnBKBUGhqhpUeaYwUACuIhku2bkcCQDGtVAiJX49IWEVvNV1IgA04eqR0dTs8SIdXQlQyVA9jVliSysAkrIwBqjzPrJFlWBzywZwAD5nqBcrN2ezLexbV7e8Ma-tnSLvUGcwD2BgwNjaYgAYjZUNIMIIAAqaWo6YEAHnCPXG6gBcBEAA99rIACboAG7fZQQ5tY4kAB8Aw0jWBlLBELg0NhwHhACURKjYQYMTiklcCUTSXBySCqUdAvTGeo0T06QAKCYVOBgZGo2SmTlwxE69FYqWiqB07Ga9QQKSoEz+Y7oMparoYYD4h2wqDAnQAbi0ugAMiJkpovTVfXAnBUAJSmRIQYDEv2NMgytlQmGG3awuBo3HikQksnmikHOVyBnlDSHS5QHk2ADupgBqYqXQFlNMjQwtVkNMCmGz3MEwo0hcJxclXx+8z+APpqttMCiORmvy6PXjHJH8NzMAL5rpJD7A+Ow658PHIvreKnJee3w35qXK7Xr23+r3ggPR-rJ5noOciXoaN5mneRaPrEi5qu+pgtF4AS+DuBqjn+AKAf2wEnGCV5jhqk4SugbAPAYb52g6X67vhGHHqe2EXnhYFEdOpb1uWso4WqGrvlRyFBLciF9EJMwieueSbt0AnpKhP50RxMrUsc1YatY8B1is7YaNycCqgAhO+saKpqmlXEopFtF0Nj7ACqqxhqDh4pohxwLpqoAESRB5bknO+IShPpnm8D5wJwEZxk1taZlQKcsRWbYtnmsudoxJEDkVE5IguWIbkYHpXkzKFfl2gFcBBYVRDFeFdqxpFGoVDFcUzDECU2f8yUZRoWU5eo7mVRw1X+cEoQDagQ21SZjXms1eStdZSULB19YpauaWeF16g9a5UWmTNFnUPNiXLSs9mOc4HYiDAnQnLt031o2LZwDFVrWp2wCClAPaZRqjhAA)

I'm super busy with work right now and have to switch back to that for the time being, but I'll keep working on this tonight if I can 😄

Comment (Ethan-Arrowood) [reply test/types.test.ts:47]: As I continue working on my playground example I wanted to share something I'm pretty certain about.

One of the challenges here is that when you create a fastify plugin there is no way to declare that it is for any specific server type or set of server types. This is even a limitation in JavaScript; the logic for if a plugin is locked into http, https, or http2 would have to exist within the plugin code itself (i.e. some top level check for the right server type). This mean in TypeScript land we can only safely type a fastify plugin as "it can be used for any server type" too. A TS dev can utilize type guarding and casting within their function if they need to utilize something specific about the fastify instance, just like a JS dev would need to check for the existence of such properties or methods. 

What I can derive from this is maybe we consider adding some new options to the fastify-plugin thing that allows us to specify what server type a plugin should be for; this would then enable the TS side to use discriminant unions and overloads to make plugins type safe based on their options object.  

Your idea @thomheymann might be a great option. I'm going to try this out and see what the dev experience is like.
> When authoring a plugin, we will not know what generics will be passed into FastifyInstance so maybe the types need to cater for that?
e.g. FastifyPlugin<Server = http.Server & https.Server & http2.Http2Server>

Comment (mcollina) [reply test/types.test.ts:47]: My point of view is that developers should not have to specify any generics for plugins, and cast things if they need to. Essentially, fastify-plugin should be pass-through.

Comment (Ethan-Arrowood) [reply test/types.test.ts:47]: [Here](https://www.typescriptlang.org/play/index.html?ssl=126&ssc=1&pln=134&pc=1#code/JYWwDg9gTgLgBACxjMBnOAzKERwORIqp4BQoksiyYm2uB1p508hYATLTvm+6STACeYAKZwA8mBjAIAO3QBeOAG8ScdXAA2EAOY6RUAPwAuOACMIETSICGskgF8SA4WIAS1VJOlzFEqTLycABkKmoabKimkQB0AMoGAG4G3oGojs5ConAeKOypvnBKBUGhqhpUeaYwUACuIhku2bkcCQDGtVAiJX49IWEVvNV1IgA04eqR0dTs8SIdXQlQyVA9jVliSysAkrIwBqjzPrJFlWBzywZwAD5nqBcrN2ezLexbV7e8Ma-tnSLvUGcwD2BgwNjaYgAYjZUNIMIIAAqaWo6YEAHnCPXG6gBcBEAA99rIACboAG7fZQQ5tY6nNgPAwkAB8Aw0AAoJhU4MDYXYIaZobDgPDdrC0QCmdiueoIFIov5jqgpdLZASYKY2QBKIosxIQYDEzma0x6g0AbkaGFqshpgTgkIRGIqWM5uLVIhJZKSBgpByOgWZHIqGFkAphcMRyNRsjRWLgEuVcFlMFQJgVaUTqsJaa1OrgpsNxvt4eFkZR6LjEtZ6i6ME6JxDjWBlLBEOLQvhACURKjYQYnfHvVA8YSPaTB5coL6qf65HTqAzASzyhpYwFfEyg9K4GAo8Cwx2y9G14rRhOVpLORVk6nTH0yjuuhhgPi07CoMCdGatLoADIiZJNDfGpPzgJwKiLAsLScMgQSgVsoRLEV5BgcUhxHIlx3JODqWOZdOUOSdOxsAB3UwAQtCoul7SkDwjbsaP7CVKI0Xdy3kOjSyRdi0TsQQzwlABtABdaDnCtG1aTBQ8BxxdD3U9Z5vhmX5FiHQMb1MH55j+Hoi0FCNRVQiUSAk2052kiNZPPK4FPHL5tMnDS5S0mY9M45CxRMsypKQwRrLdUdFNiAFnJTVyiHc9tDJQtDJyZUzrXMhs-IC+SgvshdQs3G80yigzSyMuKL0SyS7Us0titsjKvUnadcMCTdOVyu91yCW4Wi8Nr0A6tzuqebSFm6bqSH0vyiuwykGrkfCKmseBCJWFj1GFOA2QAQhvbUV2lRariUek2i6Gx9gBLVOQcPFNEObkMDWgAiSJ7u5E4bxCUJ1rZR6Zme4EkzlTVtqvDQ9uHA7PBiI7bFOoc2RvGJIk1C6rpu26Ht4X7Xrld64E+76iEx-6U0B6suVB+c8kh46Ycnc6KkukRrrENGvoxl6ifQYJQnxtBCa20mKnJ8HKahk7-h0tTafhxHkcZm6drJ9DhfOUWaZWOmNBgio2l8eA2OjeUCvhbjjz4gShyZETThE5xtd1uBqOAPsoA8wQGKdykqqgFklDZfXgW1BRZu3f35BiMBalQBA-b3WQkfp22NFresBZBodiLIuBQcTKiew9gwc9Y2P0gTmCdZQncAEZXZN9FlAcM96SrX2eRgM8bzPLMYED5cnHL2Ed3YGvY7RevG4h5u1tb9u5U7tUe5UPv7bAABmYeeLHpTHIvU42Wnjm58JBf65Ifu9YAFnX49N4clSJfF+Ld-3ju4C74+y-thFK9OB0-crpGz5wARJwJQv8OAAM-ivH+CI-YrwgRXBE59oF+3PkjU+9s2C4iUBVeE51MFDhiI7Z2bIv5I3wZOQhediHALIdQAElDGJQBIXAkg5CVgMPzkwxBaDAGRCwZgPybJlB3FMPXMCtCiD0KIZSEh-9WGeCkVQmRND5GSIIdIgwzCJFoEUYwkhqDnC8LvpOU4ODBBCOeMMeo4jVGtHUUozRpDbFvHsXolRvBdGcK0c4zx1CDHoIrh4++-CzEWKGHAGo9Rx5EFEZdBw2iXFDV8couRQSkmuK8e4u+6SKEaK4SwtJfxkmOIMUAA) is my latest playground link. 

This is a tough issue; I'm having a hard time landing on some resolution. **What do you feel like is the best developer experience for TypeScript developers here?** I need more opinions so we can zero-in on a solution

Comment (Ethan-Arrowood) [reply test/types.test.ts:47]: Based on @mcollina answer I believe the `FastifyPlugin` interface in the `fastify` repo should remove the generics `RawServer`, `RawRequest`, and `RawResponse`. Also removing them from this repo.

If the user needs a specific server instance in their plugin they need to use type casting or type guarding

Comment (mcollina) [reply test/types.test.ts:47]: Can we try that?

Comment (Ethan-Arrowood) [reply test/types.test.ts:47]: @thomheymann do you want to try making the changes I recommended?

Comment (thomheymann) [reply test/types.test.ts:47]: Would it be possible to just leave the `FastifyPlugin` instance type as the `ServerIntersection` by default? It reflects the reality of plugin authors not knowing what exact server type will be used. For those plugins that depend on a specific server or modified request/response objects those should be able to be passed in but most plugins will be generic. 

I've had a stab here at this here:
https://www.typescriptlang.org/play/index.html?ssl=113&ssc=2&pln=108&pc=1#

* I mocked http/https/http2 namespaces as the playground couldn't find dependencies
* `FastifyPlugin` instance type is now union of all server types (its called intersection but it's a union, worst naming in typescript ever):

    ```ts
    interface FastifyPlugin<Options, Server extends ServerIntersection = ServerIntersection> {
        (
            instance: FastifyInst<ServerIntersection>,
            opts: Options,
            next: () => void
        ): void;
    }
    ```

* Plugin authors get correct instance type:

    ```ts
    // Generic plugin
    const p4: FastifyPlugin<{}> = (inst, opts, next) => {
        console.log(inst.serverRaw.type) // "http" | "https" | "http2" | "https2"
    }

    // Specific plugin
    const p3: FastifyPlugin<{}, https.Server> = (inst, opts, next) => {
        console.log(inst.serverRaw.type) // "https"
    }
    ```

* Userland Fastify instance type is correctly inferred based on options
* Userland plugin registration works without typing errors

Comment (mcollina) [reply test/types.test.ts:47]: Seems a good idea to me but my TS is far from good.

Comment (Ethan-Arrowood) [reply test/types.test.ts:47]: Yeah this looks good. I'm happy shipping this now and if things change in the future we can of course change it ourselves.

Yeah this is one of those edge cases that is really hard to solve correctly because it wouldn't happen when using TS from the ground up. If I have the opportunity I'll pass this along to the TS team for consideration.

Comment (thomheymann) [reply test/types.test.ts]: This PR has been updated but will fail until corresponding Fastify PR has been merged in:
https://github.com/fastify/fastify/pull/2183

**Comment by mcollina** [comment]
@Ethan-Arrowood can you take a look?

**Comment by thomheymann** [comment]
Need to get this one in first to fix build - Travis is bogged:
https://github.com/fastify/fastify-plugin/pull/86

**Comment by thomheymann** [comment]
@mcollina This isn't going to work in Node 6 unless Fastify is transpiled and published (atm we're pointing to `next` branch)

```
/home/runner/work/fastify-plugin/fastify-plugin/node_modules/fastify/lib/reply.js:173
    ...this.raw.getHeaders(),
    ^^^

SyntaxError: Unexpected token ...
    at Object.Module._extensions..js (module.js:586:10)
    at Object.Module._extensions..js (module.js:586:10)
```

**Comment by mcollina** [comment]
This branch should target node 10+, we 'll release it as a major.

**Comment by thomheymann** [comment]
ok this can be merged in once Fastify v3 has been released

outstanding tasks:
update package.json to Fastify v3
